distributed/shuffle/tests/test_shuffle.py

```diff
@@ -469,3 +476,208 @@ async def test_clean_after_close(c, s, a, b):
 
     # clean_scheduler(s)  # TODO
     # clean_worker(b)  # TODO
+
+
+class PooledRPCShuffle(PooledRPCCall):
+    def __init__(self, shuffle: Shuffle):
+        self.shuffle = shuffle
+
+    def __getattr__(self, key):
+        async def _(**kwargs):
+            from distributed.protocol.serialize import nested_deserialize
+
+            method_name = key.replace("shuffle_", "")
+            kwargs.pop("shuffle_id", None)
+            # TODO: This is a bit awkward. At some point the arguments are
+            # already getting wrapped with a `Serialize`. We only want to unwrap
+            # here.
+            kwargs = nested_deserialize(kwargs)
+            meth = getattr(self.shuffle, method_name)
+            return await meth(**kwargs)
+
+        return _
+
+
+class ShuffleTestPool:
+    def __init__(self, *args, **kwargs):
+        self.shuffles = {}
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, addr: str, *args: Any, **kwargs: Any) -> PooledRPCShuffle:
+        return PooledRPCShuffle(self.shuffles[addr])
+
+    async def fake_broadcast(self, msg):
+
+        op = msg.pop("op").replace("shuffle_", "")
+        out = {}
+        for addr, s in self.shuffles.items():
+            out[addr] = await getattr(s, op)()
+        return out
+
+    def new_shuffle(
+        self, name, worker_for_mapping, schema, directory, loop, Shuffle=Shuffle
+    ):
+        s = Shuffle(
+            column="_partition",
+            worker_for=worker_for_mapping,
+            # FIXME: Is output_workers redundant with worker_for?
+            output_workers=set(worker_for_mapping.values()),
+            schema=schema,
+            directory=directory / name,
+            id=ShuffleId(name),
+            local_address=name,
+            nthreads=2,
+            rpc=self,
+            broadcast=self.fake_broadcast,
+            memory_limiter_disk=ResourceLimiter(10000000),
+            memory_limiter_comms=ResourceLimiter(10000000),
+        )
+        self.shuffles[name] = s
+        return s
+
+
+# 36 parametrizations
+# Runtime each ~0.1s
+@pytest.mark.parametrize("n_workers", [1, 10])
+@pytest.mark.parametrize("n_input_partitions", [1, 2, 10])
+@pytest.mark.parametrize("npartitions", [1, 20])
+@pytest.mark.parametrize("barrier_first_worker", [True, False])
+@gen_test(timeout=5)
+async def test_basic_lowlevel_shuffle(
+    tmpdir,
+    loop_in_thread,
+    n_workers,
+    n_input_partitions,
+    npartitions,
+    barrier_first_worker,
+):
+    dfs = []
+    rows_per_df = 10
+    for ix in range(n_input_partitions):
+        df = pd.DataFrame({"x": range(rows_per_df * ix, rows_per_df * (ix + 1))})
+        df["_partition"] = df.x % npartitions
+        dfs.append(df)
+
+    workers = list("abcdefghijklmn")[:n_workers]
+
+    worker_for_mapping = {}
+
+    for part in range(npartitions):
```

Comment:
nit for readability:

```suggestion
    for output_partition in range(npartitions):
```
(requires similar adjustments below)